evm/utils/blobs.py

```diff
@@ -0,0 +1,68 @@
+from itertools import (
+    chain,
+)
+
+from eth_utils import (
+    keccak,
+)
+
+from evm.constants import (
+    CHUNK_SIZE,
+    COLLATION_SIZE,
+)
+
+
+def chunk_iterator(blob):
+    if len(blob) % CHUNK_SIZE != 0:
+        raise ValueError("Blob size is {} which is not a multiple of chunk size ({})".format(
+            len(blob),
+            CHUNK_SIZE,
+        ))
+    for chunk_start in range(0, len(blob), CHUNK_SIZE):
+        yield blob[chunk_start:chunk_start + CHUNK_SIZE]
+
+
+def calc_merkle_root(leaves):
```

Comment:
I can't help myself when I see something like this which can be done functionally.  What do you think about the following?

1. do up-front validation that `is_power_of_2(len(leaves)) == True`.
2. calc `p` such that `2 ** p == len(leaves)`
3. run a non-loop based reduction using a fixed number of `cytoolz.partition` based steps.

```python
import math
import itertools
from cytoolz import pipe, compose

@to_tuple
def hash_layer(layer):
    for left, right in partition(2, layer):
        yield keccak(left + right)

def calc_merkle_root(leaves):
    p = math.log2(len(leaves))
    if p != int(p):
        # not a power of 2
    root, *extras = pipe(leaves, *itertools.repeat(hash_layer, int(p)))
    if extras:
        raise Exception("Invariant: should only be a single value")
    return root
```

Reply:
I like this, the only issue is that it wouldn't work with iterators (because the number of leaves is not known in advance). I tried to make use of iterators for the sake of efficiency as calculating the Merkle root of a full-sized block already takes almost a second on my machine. But blobs are only a megabyte in size and iterators should only decrease memory usage, so that argument is flawed. So I'm gonna implement your suggestion and work with lists.

Reply:
In *theory* this can be made to work with iterators, but you'll likely have to modify the implementation to be recursive which is *probably* fine, but could be problematic at high depths which should be easy enough to validate.

Either way is fine with me.